File: ui/src/utils/insights.ts
import type { StepSummary, TraceInsights, TraceSummary } from '../types';

export function computeInsights(trace: TraceSummary): TraceInsights {
  const durations = trace.steps.map((step) => ({
    stepId: step.id,
    name: step.name,
    durationMs: step.durationMs ?? 0,
  }));

  const topLatencySteps = durations.sort((a, b) => b.durationMs - a.durationMs).slice(0, 3);

  const costByType: Record<string, number> = {};
  const costByTool: Record<string, number> = {};
  for (const step of trace.steps) {
    if (step.metrics?.costUsd != null) {
      costByType[step.type] = (costByType[step.type] ?? 0) + step.metrics.costUsd;
      if (step.type === 'tool_call') {
        costByTool[step.name] = (costByTool[step.name] ?? 0) + step.metrics.costUsd;
      }
    }
  }

  const errors = trace.steps.filter((step) => step.status === 'failed').length;
  const retries = trace.steps.filter((step) => Boolean(step.retryOfStepId)).length;

  const wallTimeMs = trace.metadata.wallTimeMs;
  const workTimeMs = trace.metadata.workTimeMs ?? trace.steps.reduce((sum, step) => sum + (step.durationMs ?? 0), 0);

  const timing = computeTimingHealth(trace);
  const ioWarnings = computeIoWarnings(trace.steps);
  const criticalPathMs = computeCriticalPath(trace.steps);
  const concurrency = computeConcurrency(trace);
  const retryPatterns = computeRetryPatterns(trace.steps);

  return {
    topLatencySteps,
    costByType,
    costByTool,
    costByModel: { [trace.metadata.modelId]: trace.metadata.totalCostUsd ?? 0 },
    errors,
    retries,
    wallTimeMs,
    workTimeMs,
    timing,
    ioWarnings,
    criticalPathMs,
    concurrency,
    retryPatterns,
  };
}

export function stepCostLabel(step: StepSummary) {
  const cost = step.metrics?.costUsd;
  if (cost == null) return null;
  return `$${cost.toFixed(3)}`;
}

function parseIso(value?: string | null) {
  if (!value) return null;
  const parsed = Date.parse(value);
  return Number.isNaN(parsed) ? null : parsed;
}

function computeTimingHealth(trace: TraceSummary) {
  const traceStart = parseIso(trace.startedAt);
  const traceEnd = parseIso(trace.endedAt ?? undefined);
  const missingStepIds: string[] = [];
  const skewedStepIds: string[] = [];
  const issues: string[] = [];

  if (traceStart == null) {
    issues.push('Trace start time missing or invalid.');
  }

  trace.steps.forEach((step) => {
    const start = parseIso(step.startedAt);
    const end = parseIso(step.endedAt ?? step.startedAt);
    if (start == null || end == null) {
      missingStepIds.push(step.id);
      return;
    }
    if (end < start) {
      skewedStepIds.push(step.id);
      return;
    }
    if (traceStart != null && start < traceStart) {
      skewedStepIds.push(step.id);
    }
    if (traceEnd != null && end > traceEnd) {
      skewedStepIds.push(step.id);
    }
  });

  if (missingStepIds.length) issues.push(`Missing timestamps on ${missingStepIds.length} steps.`);
  if (skewedStepIds.length) issues.push(`Timestamp skew detected on ${skewedStepIds.length} steps.`);

  return {
    degraded: Boolean(missingStepIds.length || skewedStepIds.length || traceStart == null),
    issues,
    missingStepIds,
    skewedStepIds,
  };
}

function computeIoWarnings(steps: StepSummary[]) {
  const toolSteps = new Map<string, StepSummary>();
  steps.forEach((step) => {
    if (step.type === 'tool_call' && step.toolCallId) {
      toolSteps.set(step.toolCallId, step);
    }
  });

  const emitted = new Set<string>();
  const consumed = new Set<string>();
  const warnings: Array<{ kind: string; message: string; stepId?: string; toolCallId?: string }> = [];

  steps.forEach((step) => {
    if (step.type !== 'llm_call' || !step.io) return;
    (step.io.emittedToolCallIds ?? []).forEach((callId) => {
      emitted.add(callId);
      if (!toolSteps.has(callId)) {
        warnings.push({
          kind: 'missing_tool_step',
          message: `Emitted toolCallId ${callId} has no tool step.`,
          stepId: step.id,
          toolCallId: callId,
        });
      }
    });
    (step.io.consumedToolCallIds ?? []).forEach((callId) => {
      consumed.add(callId);
      if (!toolSteps.has(callId)) {
        warnings.push({
          kind: 'missing_tool_step',
          message: `Consumed toolCallId ${callId} has no tool step.`,
          stepId: step.id,
          toolCallId: callId,
        });
      }
      if (!emitted.has(callId)) {
        warnings.push({
          kind: 'consume_without_emit',
          message: `Consumed toolCallId ${callId} without emission.`,
          stepId: step.id,
          toolCallId: callId,
        });
      }
    });
  });

  toolSteps.forEach((step, callId) => {
    if (!emitted.has(callId)) {
      warnings.push({
        kind: 'unemitted_tool',
        message: `Tool step ${step.id} has toolCallId ${callId} never emitted.`,
        stepId: step.id,
        toolCallId: callId,
      });
    }
    if (!consumed.has(callId)) {
      warnings.push({
        kind: 'unconsumed_tool',
        message: `Tool step ${step.id} has toolCallId ${callId} never consumed.`,
        stepId: step.id,
        toolCallId: callId,
      });
    }
  });

  return warnings;
}

function computeCriticalPath(steps: StepSummary[]) {
  const durations = new Map<string, number>();
  const children = new Map<string, string[]>();
  const roots: string[] = [];
  steps.forEach((step) => {
    durations.set(step.id, step.durationMs ?? 0);
    if (step.parentStepId) {
      const list = children.get(step.parentStepId) ?? [];
      list.push(step.id);
      children.set(step.parentStepId, list);
    } else {
      roots.push(step.id);
    }
  });

  const memo = new Map<string, number>();
  const visit = (stepId: string): number => {
    if (memo.has(stepId)) return memo.get(stepId) ?? 0;
    const childIds = children.get(stepId) ?? [];
    const longestChild = childIds.reduce((max, childId) => Math.max(max, visit(childId)), 0);
    const value = (durations.get(stepId) ?? 0) + longestChild;
    memo.set(stepId, value);
    return value;
  };

  if (!roots.length) return 0;
  return Math.max(...roots.map((rootId) => visit(rootId)));
}

function computeConcurrency(trace: TraceSummary) {
  const start = parseIso(trace.startedAt);
  if (start == null) return { buckets: [], peak: 0 };
  const end =
    parseIso(trace.endedAt ?? undefined) ??
    Math.max(
      ...trace.steps.map((step) => parseIso(step.endedAt ?? step.startedAt) ?? start),
      start
    );
  const wallMs = Math.max(1, end - start);
  const bucketCount = 12;
  const bucketMs = Math.max(1, Math.floor(wallMs / bucketCount));
  const buckets = [];
  let peak = 0;

  for (let i = 0; i < bucketCount; i += 1) {
    const bucketStart = start + i * bucketMs;
    const bucketEnd = bucketStart + bucketMs;
    let active = 0;
    trace.steps.forEach((step) => {
      const stepStart = parseIso(step.startedAt);
      const stepEnd = parseIso(step.endedAt ?? step.startedAt);
      if (stepStart == null || stepEnd == null) return;
      if (stepEnd >= bucketStart && stepStart <= bucketEnd) active += 1;
    });
    peak = Math.max(peak, active);
    buckets.push({ startMs: i * bucketMs, endMs: (i + 1) * bucketMs, active });
  }

  return { buckets, peak };
}

function computeRetryPatterns(steps: StepSummary[]) {
  const retryCounts: Record<string, number> = {};
  steps.forEach((step) => {
    if (!step.retryOfStepId) return;
    retryCounts[step.retryOfStepId] = (retryCounts[step.retryOfStepId] ?? 0) + 1;
  });
  const totalRetries = Object.values(retryCounts).reduce((sum, count) => sum + count, 0);
  const entries = Object.entries(retryCounts).sort((a, b) => b[1] - a[1]).slice(0, 3);
  return {
    totalRetries,
    retryRate: steps.length ? totalRetries / steps.length : 0,
    topRetries: entries.map(([stepId, count]) => ({ stepId, count })),
  };
}

File: server/trace/insights.py
from __future__ import annotations

from typing import Any, Dict, List, Optional, Tuple

from .schema import StepSummary, TraceSummary


def compute_insights(trace: TraceSummary) -> Dict[str, Any]:
    durations = _durations(trace.steps)
    top_latency = sorted(durations, key=lambda item: item["durationMs"], reverse=True)[:3]
    cost_by_type = _cost_by_type(trace.steps)
    cost_by_tool = _cost_by_tool(trace.steps)
    cost_by_model = _cost_by_model(trace)

    error_count = sum(1 for step in trace.steps if step.status == "failed")
    retry_count = sum(1 for step in trace.steps if step.retryOfStepId)

    wall_time_ms = trace.metadata.wallTimeMs
    work_time_ms = trace.metadata.workTimeMs or sum(item["durationMs"] for item in durations)
    timing = _timing_health(trace)
    io_warnings = _io_warnings(trace.steps)
    critical_path_ms = _critical_path_ms(trace.steps)
    concurrency = _concurrency_heatmap(trace, bucket_count=12)
    retry_patterns = _retry_patterns(trace.steps)

    return {
        "topLatencySteps": top_latency,
        "costByType": cost_by_type,
        "costByTool": cost_by_tool,
        "costByModel": cost_by_model,
        "errors": error_count,
        "retries": retry_count,
        "wallTimeMs": wall_time_ms,
        "workTimeMs": work_time_ms,
        "timing": timing,
        "ioWarnings": io_warnings,
        "criticalPathMs": critical_path_ms,
        "concurrency": concurrency,
        "retryPatterns": retry_patterns,
    }


def _durations(steps: List[StepSummary]) -> List[Dict[str, Any]]:
    result = []
    for step in steps:
        duration = step.durationMs
        if duration is None:
            duration = _fallback_duration(step.startedAt, step.endedAt)
        result.append({"stepId": step.id, "name": step.name, "durationMs": duration or 0})
    return result


def _cost_by_type(steps: List[StepSummary]) -> Dict[str, float]:
    totals: Dict[str, float] = {}
    for step in steps:
        if step.metrics and step.metrics.costUsd is not None:
            totals[step.type] = totals.get(step.type, 0.0) + float(step.metrics.costUsd)
    return totals


def _cost_by_tool(steps: List[StepSummary]) -> Dict[str, float]:
    totals: Dict[str, float] = {}
    for step in steps:
        if step.type != "tool_call":
            continue
        if step.metrics and step.metrics.costUsd is not None:
            totals[step.name] = totals.get(step.name, 0.0) + float(step.metrics.costUsd)
    return totals


def _cost_by_model(trace: TraceSummary) -> Dict[str, float]:
    total = trace.metadata.totalCostUsd or 0.0
    return {trace.metadata.modelId: float(total)}


def _fallback_duration(started_at: str, ended_at: str | None) -> int:
    if not started_at:
        return 0
    from datetime import datetime

    fmt = "%Y-%m-%dT%H:%M:%S.%fZ"
    try:
        start = datetime.strptime(started_at, fmt)
        end = datetime.strptime(ended_at or started_at, fmt)
    except ValueError:
        return 0
    return int((end - start).total_seconds() * 1000)


def _parse_iso(ts: Optional[str]) -> Optional["datetime"]:
    if not ts:
        return None
    from datetime import datetime

    try:
        if ts.endswith("Z"):
            return datetime.fromisoformat(ts.replace("Z", "+00:00"))
        return datetime.fromisoformat(ts)
    except ValueError:
        return None


def _timing_health(trace: TraceSummary) -> Dict[str, Any]:
    trace_start = _parse_iso(trace.startedAt)
    trace_end = _parse_iso(trace.endedAt) if trace.endedAt else None

    missing_steps: List[str] = []
    skewed_steps: List[str] = []
    issues: List[str] = []

    if not trace_start:
        issues.append("Trace start time missing or invalid.")

    for step in trace.steps:
        start = _parse_iso(step.startedAt)
        end = _parse_iso(step.endedAt) if step.endedAt else start
        if not start:
            missing_steps.append(step.id)
            continue
        if not end:
            missing_steps.append(step.id)
            continue
        if end < start:
            skewed_steps.append(step.id)
            continue
        if trace_start and start < trace_start:
            skewed_steps.append(step.id)
        if trace_end and end > trace_end:
            skewed_steps.append(step.id)

    if missing_steps:
        issues.append(f"Missing timestamps on {len(missing_steps)} steps.")
    if skewed_steps:
        issues.append(f"Timestamp skew detected on {len(skewed_steps)} steps.")

    return {
        "degraded": bool(missing_steps or skewed_steps or (not trace_start)),
        "issues": issues,
        "missingStepIds": missing_steps,
        "skewedStepIds": skewed_steps,
    }


def _io_warnings(steps: List[StepSummary]) -> List[Dict[str, Any]]:
    tool_step_by_call_id = {step.toolCallId: step for step in steps if step.type == "tool_call" and step.toolCallId}
    emitted_ids = set()
    consumed_ids = set()

    warnings: List[Dict[str, Any]] = []

    for step in steps:
        if step.type != "llm_call" or not step.io:
            continue
        for call_id in step.io.emittedToolCallIds:
            emitted_ids.add(call_id)
            if call_id not in tool_step_by_call_id:
                warnings.append(
                    {
                        "kind": "missing_tool_step",
                        "message": f"Emitted toolCallId {call_id} has no tool step.",
                        "stepId": step.id,
                        "toolCallId": call_id,
                    }
                )
        for call_id in step.io.consumedToolCallIds:
            consumed_ids.add(call_id)
            if call_id not in tool_step_by_call_id:
                warnings.append(
                    {
                        "kind": "missing_tool_step",
                        "message": f"Consumed toolCallId {call_id} has no tool step.",
                        "stepId": step.id,
                        "toolCallId": call_id,
                    }
                )
            if call_id not in emitted_ids:
                warnings.append(
                    {
                        "kind": "consume_without_emit",
                        "message": f"Consumed toolCallId {call_id} without emission.",
                        "stepId": step.id,
                        "toolCallId": call_id,
                    }
                )

    for call_id, tool_step in tool_step_by_call_id.items():
        if call_id not in emitted_ids:
            warnings.append(
                {
                    "kind": "unemitted_tool",
                    "message": f"Tool step {tool_step.id} has toolCallId {call_id} never emitted.",
                    "stepId": tool_step.id,
                    "toolCallId": call_id,
                }
            )
        if call_id not in consumed_ids:
            warnings.append(
                {
                    "kind": "unconsumed_tool",
                    "message": f"Tool step {tool_step.id} has toolCallId {call_id} never consumed.",
                    "stepId": tool_step.id,
                    "toolCallId": call_id,
                }
            )
    return warnings


def _critical_path_ms(steps: List[StepSummary]) -> int:
    durations = {step.id: step.durationMs or 0 for step in steps}
    children: Dict[str, List[str]] = {}
    roots: List[str] = []
    for step in steps:
        if step.parentStepId:
            children.setdefault(step.parentStepId, []).append(step.id)
        else:
            roots.append(step.id)

    memo: Dict[str, int] = {}

    def visit(step_id: str) -> int:
        if step_id in memo:
            return memo[step_id]
        longest_child = 0
        for child_id in children.get(step_id, []):
            longest_child = max(longest_child, visit(child_id))
        memo[step_id] = durations.get(step_id, 0) + longest_child
        return memo[step_id]

    if not roots:
        return 0
    return max(visit(root) for root in roots)


def _concurrency_heatmap(trace: TraceSummary, bucket_count: int = 12) -> Dict[str, Any]:
    start = _parse_iso(trace.startedAt)
    end = _parse_iso(trace.endedAt) if trace.endedAt else None
    if not start:
        return {"buckets": [], "peak": 0}
    if not end:
        end = start
        for step in trace.steps:
            step_end = _parse_iso(step.endedAt) or _parse_iso(step.startedAt)
            if step_end and step_end > end:
                end = step_end

    wall_ms = max(1, int((end - start).total_seconds() * 1000))
    bucket_ms = max(1, wall_ms // bucket_count)

    buckets = []
    peak = 0
    for i in range(bucket_count):
        bucket_start = start.timestamp() * 1000 + i * bucket_ms
        bucket_end = bucket_start + bucket_ms
        active = 0
        for step in trace.steps:
            step_start = _parse_iso(step.startedAt)
            step_end = _parse_iso(step.endedAt) if step.endedAt else step_start
            if not step_start or not step_end:
                continue
            step_start_ms = step_start.timestamp() * 1000
            step_end_ms = step_end.timestamp() * 1000
            if step_end_ms >= bucket_start and step_start_ms <= bucket_end:
                active += 1
        peak = max(peak, active)
        buckets.append({"startMs": i * bucket_ms, "endMs": (i + 1) * bucket_ms, "active": active})

    return {"buckets": buckets, "peak": peak}


def _retry_patterns(steps: List[StepSummary]) -> Dict[str, Any]:
    total_steps = len(steps)
    retry_of: Dict[str, int] = {}
    for step in steps:
        if not step.retryOfStepId:
            continue
        retry_of[step.retryOfStepId] = retry_of.get(step.retryOfStepId, 0) + 1

    top_retries = sorted(retry_of.items(), key=lambda item: item[1], reverse=True)[:3]
    return {
        "totalRetries": sum(retry_of.values()),
        "retryRate": (sum(retry_of.values()) / total_steps) if total_steps else 0,
        "topRetries": [{"stepId": step_id, "count": count} for step_id, count in top_retries],
    }
